src/ILCompiler.ReadyToRun/src/Compiler/DependencyAnalysis/ReadyToRunCodegenNodeFactory.cs

```diff
@@ -46,39 +64,555 @@ public sealed class ReadyToRunCodegenNodeFactory : NodeFactory
 
         public ImportSectionsTableNode ImportSectionsTable;
 
-        protected override IMethodNode CreateMethodEntrypointNode(MethodDesc method)
+        public Import ModuleImport;
+
+        public ImportSectionNode EagerImports;
+
+        public ImportSectionNode MethodImports;
+
+        public ImportSectionNode StringImports;
+
+        public ImportSectionNode HelperImports;
+
+        public ImportSectionNode PrecodeImports;
+
+        /// <summary>
+        /// TODO: this will need changing when compiling multiple files. Ideally this
+        /// should be switched every time we compile a method in a particular module
+        /// because this is what provides context for reference token resolution.
+        /// </summary>
+        EcmaModule _inputModule;
+
+        SignatureContext _signatureContext;
+
+        public SignatureContext SignatureContext
+        {
+            get
+            {
+                if (_signatureContext == null)
+                {
+                    _signatureContext = new SignatureContext(this, _inputModule);
+                }
+                return _signatureContext;
+            }
+        }
+
+
+        Dictionary<MethodDesc, IMethodNode> _methodMap = new Dictionary<MethodDesc, IMethodNode>();
+
+        public IMethodNode MethodEntrypoint(MethodDesc method, mdToken token, bool isUnboxingStub = false)
```

Comment:
What will require some amount of thinking here and in similar spots:

* Do we need to keep track of the declaring scope of the token (e.g. the EcmaModule)?
* What is the primary key used to identify method entrypoints? Is `MethodDesc` the primary key? A single method can be referenced through different tokens (even in a single file compilation - e.g. calling a varargs method). Do we maybe need to wrap the method entrypoint in a different node? One example where we do that in existing code is `ShadowConcreteMethodNode` - this node can be used everywhere where a normal `MethodEntrypoint` is used, but it captures additional information that the wrapped MethodEntrypoint doesn't have (in case of shadow concrete method, it's the exact instantiation of canonical method bodies; in case or ready to run, this could capture the token through which the method entrypoint was referenced).

Reply:
You're right of course. The current implementation is a hack good enough to work for single-file. I'll update the code by passing the module around as a next commit. After that I believe we should be ready to produce a final PR against master R2R branch.

Sadly one confusing bit is that Amy made her commits against trylek/tree/RemainingR2RChanges while this PR was sent against RemainingR2RChangesResolved. I have rebased the RemainingR2RChanges branch, resolved all conflicts, applied your feedback and reverted unwanted changes (repro/Program.cs, the general version of the field layout managers, the general MultiFileCompilationModuleGroup). Please let me know is there's a way to project these follow-up changes in this PR.